File: tst/blend_raw.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_view("Blend", -1, -1,
        &(Pico_Rel_Dim){'!', {640, 360}, NULL},
        NULL,
        &(Pico_Rel_Dim){'!', {64, 36}, NULL},
        NULL,
        NULL
    );

    {
        puts("pixel dimming");
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_pixel(&(Pico_Rel_Pos){'!', {32,18}, PICO_ANCHOR_NW, NULL});
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-01");
            }
        }
    }
    {
        puts("text dimming");
        Pico_Abs_Dim dim = pico_get_text("SOME TEXT", &(Pico_Rel_Dim){'!',{0,10},NULL});
        Pico_Rel_Rect r = {'!', {32-dim.w/2, 18-5, 0, 10}, PICO_ANCHOR_NW, NULL};
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_text("SOME TEXT", &r);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-02");
            }
        }
    }
    {
        puts("rect dimming");
        Pico_Rel_Rect r = {'!', {32-30, 18-6, 60, 12}, PICO_ANCHOR_NW, NULL};
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_rect(&r);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-03");
            }
        }
    }
    {
        puts("oval dimming");
        Pico_Rel_Rect r = {'!', {32-30, 18-6, 60, 12}, PICO_ANCHOR_NW, NULL};
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_oval(&r);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-04");
            }
        }
    }
    {
        puts("tri dimming");
        Pico_Rel_Pos p1 = {'!', { 2, 12}, PICO_ANCHOR_NW, NULL};
        Pico_Rel_Pos p2 = {'!', { 2, 24}, PICO_ANCHOR_NW, NULL};
        Pico_Rel_Pos p3 = {'!', {62, 24}, PICO_ANCHOR_NW, NULL};
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_tri(&p1, &p2, &p3);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-05");
            }
        }
    }
    {
        puts("line dimming");
        Pico_Rel_Pos p1 = {'!', {12, 12}, PICO_ANCHOR_NW, NULL};
        Pico_Rel_Pos p2 = {'!', {52, 24}, PICO_ANCHOR_NW, NULL};
        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_line(&p1, &p2);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-06");
            }
        }
    }
    {
        puts("poly dimming");
        Pico_Rel_Pos poly[] = {
            {'!', { 5,  5}, PICO_ANCHOR_NW, NULL},
            {'!', {59, 10}, PICO_ANCHOR_NW, NULL},
            {'!', {20, 31}, PICO_ANCHOR_NW, NULL},
        };

        for (int a=255; a>0; a-=5) {
            pico_output_clear();
            pico_set_alpha(a);
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_poly(3, poly);
            pico_input_delay(10);
            if (a == 120) {
                _pico_check("blend_raw-07");
            }
        }
    }

    pico_init(0);
    return 0;
}

File: lua/tst/blend_raw.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Blend", dim={'!', w=640, h=360} }
pico.set.view { dim={'!', w=64, h=36} }

do
    print("pixel dimming")
    local p = {'!', x=32, y=18, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.pixel(p)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-01")
        end
    end
end

do
    print("text dimming")
    local d = {'!', w=0, h=10}
    pico.get.text("SOME TEXT", d)
    local r = {'!', x=32-d.w//2, y=18-5, w=d.w, h=10, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.text("SOME TEXT", r)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-02")
        end
    end
end

do
    print("rect dimming")
    local r = {'!', x=32-30, y=18-6, w=60, h=12, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.rect(r)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-03")
        end
    end
end

do
    print("oval dimming")
    local r = {'!', x=32-30, y=18-6, w=60, h=12, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.oval(r)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-04")
        end
    end
end

do
    print("tri dimming")
    local p1 = {'!', x=2,  y=12, anchor='NW'}
    local p2 = {'!', x=2,  y=24, anchor='NW'}
    local p3 = {'!', x=62, y=24, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.tri(p1, p2, p3)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-05")
        end
    end
end

do
    print("line dimming")
    local p1 = {'!', x=12, y=12, anchor='NW'}
    local p2 = {'!', x=52, y=24, anchor='NW'}
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.line(p1, p2)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-06")
        end
    end
end

do
    print("poly dimming")
    local poly = {
        {'!', x=5,  y=5, anchor='NW'},
        {'!', x=59, y=10, anchor='NW'},
        {'!', x=20, y=31, anchor='NW'},
    }
    for a = 255, 1, -5 do
        pico.output.clear()
        pico.set.alpha(a)
        pico.set.color.draw('red')
        pico.output.draw.poly(poly)
        pico.input.delay(10)
        if a == 120 then
            pico.check("blend_raw-07")
        end
    end
end

pico.init(false)
